crates/connector_proxy/src/connector_runner.rs

```diff
@@ -75,14 +80,68 @@ pub async fn run_airbyte_source_connector(
     )?;
 
-    let adapted_response_stream = NetworkTunnelCaptureInterceptor::adapt_response_stream(
-        op,
-        airbyte_interceptor.adapt_response_stream(op, response_stream(child_stdout))?,
-    )?;
+    let res_stream =
+        airbyte_interceptor.adapt_response_stream(op, response_stream(child_stdout))?;
+
+    // Keep track of whether we did send a Driver Checkpoint as the final message of the response stream
+    // See the comment of the block below for why this is necessary
+    let (tp_sender, tp_receiver) = oneshot::channel::<bool>();
+    let res_stream = if *op == FlowCaptureOperation::Pull {
+        Box::pin(stream::try_unfold(
+            (false, res_stream, tp_sender),
+            |(transaction_pending, mut stream, sender)| async move {
+                let (message, raw) = match stream.next().await {
+                    Some(bytes) => {
+                        let bytes = bytes?;
+                        let mut buf = &bytes[..];
+                        let msg = decode_message::<PullResponse, _>(&mut buf)
+                            .await
+                            .unwrap()
+                            .unwrap();
+                        (msg, bytes)
+                    }
+                    None => {
+                        sender.send(transaction_pending).unwrap();
+                        return Ok(None);
+                    }
+                };
+
+                Ok(Some((raw, (!message.checkpoint.is_some(), stream, sender))))
+            },
+        ))
+    } else {
+        res_stream
+    };
+
+    let adapted_response_stream =
+        NetworkTunnelCaptureInterceptor::adapt_response_stream(op, res_stream)?;
 
     streaming_all(child_stdin, adapted_request_stream, adapted_response_stream).await?;
 
-    check_exit_status("airbyte source connector:", child.wait().await)
+    let exit_status = check_exit_status("airbyte source connector:", child.wait().await);
+
+    // There are some Airbyte connectors that write records, and exit successfully, without ever writing
+    // a state (checkpoint). In those cases, we want to provide a default empty checkpoint. It's important that
+    // this only happens if the connector exit successfully, otherwise we risk double-writing data.
+    if exit_status.is_ok() && *op == FlowCaptureOperation::Pull {
+        // the received value (transaction_pending) is true if the connector writes output messages and exits _without_ writing
+        // a final state checkpoint.
+        if tp_receiver.await.unwrap() {
+            // We generate a synthetic commit now, and the empty checkpoint means the assumed behavior
+            // of the next invocation will be "full refresh".
+            tracing::warn!("connector exited without writing a final state checkpoint, flushing the driver checkpoint.");
+            let mut resp = PullResponse::default();
+            resp.checkpoint = Some(DriverCheckpoint {
+                driver_checkpoint_json: Vec::new(),
```

Comment:
I'm thinking that it's worth changing this to return an empty json object for the checkpoint json, and to set it as a merge patch. That will ensure that a prior checkpoint doesn't get overwritten in this case.

Also, I think this block should assert that the airbyte `SyncMode` is set to `full_refresh`. If it's `incremental`, then it seems incorrect to generate a synthetic checkpoint.

Reply:
@psFried Do we have any way of being able to do merge patch in a way that works with any value? i.e. if the checkpoint from the connector is an array or a scalar, merging with `{}` will still clobber it.

Reply:
@psFried My understanding of the Airbyte protocol is that the state file is generally optional, for all sync modes. I know it seems non-sensical to have an incremental connector that doesn't have a state. Let's imagine the case where an incremental connector does exit successfully without an error, and without having sent out a State message. If we do not append an empty checkpoint, the connector and the deployment will just hang forever, right? Do we want that? 🤔

Reply:
In the case that a connector claims to support incremental, we _could_ treat it as an error if it exits without a final state message. That seems pretty reasonable to me. Afterall, how is it supposed to resume correctly without it? That said, I'm not sure whether or not we can actually expect all the airbyte connectors to emit final state messages, so it's fine to have the proxy emit that final state automatically in that case. We should probably log it as a warning, though, just so we can know if it's happening.

I don't think we need to worry about the top-level state being anything other than an object, though. The [airbyte specification](https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml#L71) requires states to be objects. So I think a merge patch of `{}` would work fine.

Reply:
@psFried Okay, we already do emit a warning, and I've already changed the code to do a merge patch instead of overriding the checkpoint.